discovery.go

```diff
@@ -0,0 +1,369 @@
+					}
+				}
+				sentPeers.mux.Unlock()
+			}()
+		case rm := <-d.rmPeerNotif:
+			delete(d.newPeerNotif, rm)
+		case pid := <-d.peerNotifier:
+			d.notifyNewPeer(pid)
+		case <-d.p.ctx.Done():
+			return
+		}
+	}
+}
+
+// Advertise advertises this node's interest in a topic to a discovery service
+func (d *discover) Advertise(topic string) {
+	if d.discovery == nil {
+		return
+	}
+
+	advertisingCtx, cancel := context.WithCancel(d.p.ctx)
+
+	if _, ok := d.advertising[topic]; ok {
+		return
+	}
+	d.advertising[topic] = cancel
+
+	go func() {
+		next, err := d.discovery.Advertise(advertisingCtx, topic)
+		if err != nil {
+			log.Warningf("bootstrap: error providing rendezvous for %s: %s", topic, err.Error())
+		}
+
+		t := time.NewTimer(next)
+		for {
+			select {
+			case <-t.C:
+				next, err = d.discovery.Advertise(advertisingCtx, topic)
+				if err != nil {
+					log.Warningf("bootstrap: error providing rendezvous for %s: %s", topic, err.Error())
+				}
+				t.Reset(next)
+			case <-advertisingCtx.Done():
+				t.Stop()
+				return
+			}
+		}
+	}()
+}
+
+// StopAdvertise stops advertising this node's interest in a topic
+func (d *discover) StopAdvertise(topic string) {
+	if d.discovery == nil {
+		return
+	}
+
+	if advertiseCancel, ok := d.advertising[topic]; ok {
+		advertiseCancel()
+		delete(d.advertising, topic)
+	}
+}
+
+// Bootstrap handles initial bootstrapping of peers for a given topic
+func (d *discover) Bootstrap(async bool, topic string, dOpts ...discovery.Option) {
+	if d.discovery == nil {
+		return
+	}
+
+	d.bootstrappedMux.RLock()
+	doneCh, bootstrapStarted := d.bootstrapped[topic]
+	d.bootstrappedMux.RUnlock()
+
+	if bootstrapStarted {
+		if !async {
+			<-doneCh
+		}
+	} else {
+		d.bootstrappedMux.Lock()
+		doneCh, bootstrapStarted := d.bootstrapped[topic]
+		if bootstrapStarted {
+			d.bootstrappedMux.Unlock()
+			if !async {
+				<-doneCh
+			}
+			return
+		}
+		ch := make(chan struct{}, 1)
+		d.bootstrapped[topic] = ch
+		d.bootstrappedMux.Unlock()
+
+		if async {
+			go func() {
+				d.handleDiscovery(d.p.ctx, topic, dOpts)
+				ch <- struct{}{}
+				close(ch)
+			}()
+		} else {
+			d.handleDiscovery(d.p.ctx, topic, dOpts)
+			ch <- struct{}{}
+			close(ch)
+		}
+	}
+}
+
+// Discover searches for additional peers interested in a given topic
+func (d *discover) Discover(topic string, opts ...discovery.Option) {
+	if d.discovery == nil {
+		return
+	}
+
+	d.discoverQ <- &discoverReq{topic, opts}
+}
+
+func (d *discover) handleDiscovery(ctx context.Context, topic string, opts []discovery.Option) {
+	discoverCtx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+
+	peerCh, err := d.discovery.FindPeers(discoverCtx, topic, opts...)
+	if err != nil {
+		log.Debugf("error finding peers for topic %s: %v", topic, err)
+		return
+	}
+
+	findPeers := make(map[peer.ID]struct{})
+
+loop:
+	for {
+		select {
+		case pi, more := <-peerCh:
+			if pi.ID == d.p.host.ID() || pi.ID == "" {
+				if more {
```

Comment:
This check should happen at the very top. That is:

```
case pi, ok := <-peerCh:
  if !ok {
    break loop
  }
...
```

`ok == false` means that the channel is closed and `pi` is invalid (well, the zero value).